relay-event-normalization/src/event.rs

```diff
@@ -894,6 +1019,210 @@ fn normalize_breakdowns(event: &mut Event, breakdowns_config: Option<&Breakdowns
     }
 }
 
+fn normalize_default_attributes(event: &mut Event, meta: &mut Meta, config: &NormalizationConfig) {
+    let event_type = infer_event_type(event);
+    event.ty = Annotated::from(event_type);
+    event.project = Annotated::from(config.project_id);
+    event.key_id = Annotated::from(config.key_id.clone());
+    event.version = Annotated::from(config.protocol_version.clone());
+    event.grouping_config = config
+        .grouping_config
+        .clone()
+        .map_or(Annotated::empty(), |x| {
+            FromValue::from_value(Annotated::<Value>::from(x))
+        });
+
+    let _ = relay_event_schema::processor::apply(&mut event.platform, |platform, _| {
+        if is_valid_platform(platform) {
+            Ok(())
+        } else {
+            Err(ProcessingAction::DeleteValueSoft)
+        }
+    });
+
+    // Default required attributes, even if they have errors
+    event.errors.get_or_insert_with(Vec::new);
+    event.id.get_or_insert_with(EventId::new);
+    event.platform.get_or_insert_with(|| "other".to_string());
+    event.logger.get_or_insert_with(String::new);
+    event.extra.get_or_insert_with(Object::new);
+    event.level.get_or_insert_with(|| match event_type {
+        EventType::Transaction => Level::Info,
+        _ => Level::Error,
+    });
+    if event.client_sdk.value().is_none() {
+        event.client_sdk.set_value(get_sdk_info(config));
+    }
+
+    if event.platform.as_str() == Some("java") {
+        if let Some(event_logger) = event.logger.value_mut().take() {
+            let shortened = shorten_logger(event_logger, meta);
+            event.logger.set_value(Some(shortened));
+        }
+    }
+}
+
+/// Returns `true` if the given platform string is a known platform identifier.
+///
+/// See [`VALID_PLATFORMS`] for a list of all known platforms.
+pub fn is_valid_platform(platform: &str) -> bool {
```

Comment:
nit: Can we remove this to the top, because it's public?

Reply:
All definitions above are also public. Where would you move it?